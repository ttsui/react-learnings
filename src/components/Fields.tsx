import React from "react";

interface FieldProps {
  label: string;
  onChange: (value: string) => void;
  value?: string | number;
}

interface FormFieldProps extends FieldProps {
  type: string;
}

const FormField: React.FC<FormFieldProps> = props => (
  <label
    style={{
      color: "#2c2c2c",
      display: "block",
      fontSize: "1.35rem",
      fontFamily: "'Arial Narrow', sans-serif, sans-serif-condensed",
      marginBottom: "1em"
    }}
  >
    <div style={{ float: "left", marginBottom: "0.25em" }}>{props.label}</div>

    <div>
      <input
        name={props.label.toLowerCase().replace(/ /g, "-")}
        type={props.type}
        value={props.value}
        onChange={event => props.onChange(event.target.value)}
        style={{
          border: "1px solid #d5d5d5",
          borderRadius: "3px",
          fontFamily: "Arial, sans-serif",
          fontSize: "18px",
          padding: "0.5em",
          width: "13em"
        }}
      />
    </div>
  </label>
);

export const PhoneField: React.FC<FieldProps> = props => <FormField {...props} type="tel" />;

export const TextField: React.FC<FieldProps> = props => <FormField {...props} type="text" />;
