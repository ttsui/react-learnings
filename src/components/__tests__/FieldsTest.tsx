import React from "react";
import { fireEvent, render } from "react-testing-library";
import { PhoneField, TextField } from "../Fields";

describe("PhoneField:", () => {
  it("renders", () => {
    const { container } = render(<PhoneField label="Name" value="(02) 1234 5678" onChange={() => {}} />);

    expect(container).toMatchSnapshot();
  });

  it("calls onChange handler", () => {
    const changeHandler = jest.fn();
    const { getByLabelText } = render(<PhoneField label="Name" onChange={changeHandler} />);

    fireEvent.change(getByLabelText("Name"), { target: { value: "(02) 1234 5678" } });

    expect(changeHandler).toHaveBeenCalledWith("(02) 1234 5678");
  });
});

describe("TextField:", () => {
  it("renders", () => {
    const { container } = render(<TextField label="Name" value="John Doe" onChange={() => {}} />);

    expect(container).toMatchSnapshot();
  });

  it("calls onChange handler", () => {
    const changeHandler = jest.fn();
    const { getByLabelText } = render(<TextField label="Name" onChange={changeHandler} />);

    fireEvent.change(getByLabelText("Name"), { target: { value: "Joan" } });

    expect(changeHandler).toHaveBeenCalledWith("Joan");
  });
});
