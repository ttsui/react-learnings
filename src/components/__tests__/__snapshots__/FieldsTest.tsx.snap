// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhoneField: renders 1`] = `
<div>
  <label
    style="color: rgb(44, 44, 44); display: block; font-size: 1.35rem; font-family: 'Arial Narrow', sans-serif, sans-serif-condensed; margin-bottom: 1em;"
  >
    <div
      style="float: left; margin-bottom: 0.25em;"
    >
      Name
    </div>
    <div>
      <input
        name="name"
        style="border: 1px solid #d5d5d5; border-radius: 3px; font-family: Arial, sans-serif; font-size: 18px; padding: 0.5em; width: 13em;"
        type="tel"
        value="(02) 1234 5678"
      />
    </div>
  </label>
</div>
`;

exports[`TextField: renders 1`] = `
<div>
  <label
    style="color: rgb(44, 44, 44); display: block; font-size: 1.35rem; font-family: 'Arial Narrow', sans-serif, sans-serif-condensed; margin-bottom: 1em;"
  >
    <div
      style="float: left; margin-bottom: 0.25em;"
    >
      Name
    </div>
    <div>
      <input
        name="name"
        style="border: 1px solid #d5d5d5; border-radius: 3px; font-family: Arial, sans-serif; font-size: 18px; padding: 0.5em; width: 13em;"
        type="text"
        value="John Doe"
      />
    </div>
  </label>
</div>
`;
